assemble/bin/accumulo-cluster

```diff
@@ -646,10 +655,170 @@ function control_services() {
+  fi
+}
+
+# Kills extra server processes that are not needed according to the
+# cluster.yaml file.  Conceptually this code is trying to reconcile the
+# following three sets of servers.
+#
+#  1. The notional goal set of servers specified by cluster.yaml
+#  2. The set of servers processes seen in zookeeper
+#  3. The set of server processes known to the accumulo-cluster script.  This
+#     is derived from pid files on hosts in set 1.
+#
+# This function attempts to find extra servers in set 2 that are not specified
+# by set 1.  When it does find extra servers it removes their zookeeper locks
+# avoiding removing locks of servers in set 3. The following are different
+# situations the code will see and handle.
+#
+#  * When a host is not cluster.yaml but has some processes listed in
+#    zookeeper.  For this case all of the process with that host can be killed.
+#  * When a resource group is not in cluster.yaml but has some processes listed
+#    in zookeeper.  For this case all of the processes with that resource group
+#    can be killed.
+#  * When a host is in cluster.yaml with a target of 3 processes but has 6
+#    processes listed in zookeeper.  For this case want to kill 3 processes that
+#    do not have pid files on the host.
+#
+function prune() {
+  if [[ $ARG_LOCAL == 1 ]]; then
+    # Currently the code is structured to remove all extra servers in a single resource group.  Finer granularity is not supported.
+    echo "$(red ERROR): Prune does not support running locally"
+    exit 1
+  fi
+
+  if [[ -z ${AC_TMP_DIR+x} ]]; then
+    echo "AC_TMP_DIR is not set"
+    exit 1
+  fi
+  local service_json="$AC_TMP_DIR/accumulo-service.json"
+  "$accumulo_cmd" admin serviceStatus --json >"$service_json" 2>/dev/null || exit 1
+
+  local var_name
+  local hosts
+  declare -a groups
+
+  local manager
+  if [[ $ARG_ALL == 1 || $ARG_MANAGER == 1 ]]; then
+    prune_group "manager" "default" "1" "$MANAGER_HOSTS"
+  fi
+
+  if [[ $ARG_ALL == 1 || $ARG_GC == 1 ]]; then
+    prune_group "gc" "default" "1" "$GC_HOSTS"
+  fi
+
+  if [[ $ARG_ALL == 1 || $ARG_MONITOR == 1 ]]; then
+    prune_group "monitor" "default" "1" "$MONITOR_HOSTS"
+  fi
+
+  if [[ $ARG_ALL == 1 || $ARG_COORDINATOR == 1 ]]; then
+    prune_group "compaction-coordinator" "default" "1" "$COORDINATOR_HOSTS"
+  fi
+
+  if [[ $ARG_ALL == 1 || $ARG_TSERVER == 1 ]]; then
+    #TODO in main need to adapt to having RGs for tservers
+    prune_group "tserver" "default" "$TSERVERS_PER_HOST_default" "$TSERVER_HOSTS_default"
+  fi
+
+  if [[ $ARG_ALL == 1 || $ARG_SSERVER == 1 ]]; then
+    groups=()
+    if [[ -n $ARG_SSERVER_GROUP ]]; then
+      read -r -a groups <<<"$ARG_SSERVER_GROUP"
+    else
+      # find all groups known in zookeeper, this will allow pruning entire groups that do not even exists in cluster.yaml
+      readarray -t groups < <(jq -r ".summaries.S_SERVER.resourceGroups | .[] " "$service_json")
```

Comment:
In `parse_args` we check that `getopt` exists to fail-fast. We might want to do the same thing here in `prune`, do a check at the start of the method that `jq` exists and fail with a helpful message if not.

Reply:
updated in bddbeed